### PR TITLE
runtimevar/filevar+etcdvar: don't subscribe for changes on the first Watch call

### PR DESCRIPTION
**Pull request by vangent**

### CODE REVIEW

File: runtimevar/filevar/filevar.go [hunk 1]
```diff
@@ -109,17 +109,20 @@ func (w *watcher) WatchVariable(ctx context.Context, prevVersion interface{}, pr
...
-	if notifierErr == nil {
-		defer func() {
-			_ = w.notifier.Remove(w.file)
-		}()
+	// We may never use the notifier if we read the file below and detect a
+	// change, but we must subscribe here to avoid race conditions.
+	var notifierErr error
+	if prevVersion != nil {
+		notifierErr := w.notifier.Add(w.file)
+		if notifierErr == nil {
```
Comment (ijt) [inline runtimevar/filevar/filevar.go]: Why isn't the case of notifierErr != nil handled?

File: runtimevar/etcdvar/etcdvar.go [hunk 2]
```diff
@@ -58,7 +58,10 @@ func (w *watcher) WatchVariable(ctx context.Context, prevVersion interface{}, pr
 
 	// Create a watching channel in case the variable hasn't changed.
 	// We must create it now before the Get to avoid race conditions.
-	ch := w.client.Watch(ctx, w.name)
+	var ch clientv3.WatchChan
```
Comment (ijt) [inline runtimevar/etcdvar/etcdvar.go]: Shouldn't there be an additional conformance test for the changes in this pull request?

**Comment by ijt** [comment]
What is the motivation for this change?
Performance?
Correctness?
What goes wrong without it?

**Comment by vangent** [comment]
> What is the motivation for this change?
> Performance?
> Correctness?
> What goes wrong without it?

Performance. It doesn't make sense to initialize a subscriber/notifier if you know in advance you're not going to use it.

File: runtimevar/etcdvar/etcdvar.go (hunk 2 above)
Comment (vangent) [reply runtimevar/etcdvar/etcdvar.go]: No, there's no change in behavior, we're just avoiding some unnecessary work.

File: runtimevar/filevar/filevar.go (hunk 1 above)
Comment (vangent) [reply runtimevar/filevar/filevar.go]: It is handled down below, if we end up needing the notifier, see line 147.

**Comment by ijt** [comment]
So should there be a benchmark to show that it improves performance?

**Comment by vangent** [comment]
> So should there be a benchmark to show that it improves performance?

IMHO that's not necessary. We don't have a benchmark today, and it might be hard to show improvement anyway (how long does it take to call fsnotify?). The etcd one saves an RPC, but since the test runs against a local instance, it probably doesn't take long either.

**Comment by ijt** [comment]
You could add the benchmark in a separate commit, then reorder the commits
(use git rebase -i) to put the benchmark before this change, then run the
benchmark with and without the change. Otherwise we don't know if it makes
the code run any faster.

**Comment by vangent** [comment]
I don't think it will make the code noticeably faster running locally. I think it's still good to avoid unnecessary operations though.

**Comment by ijt** [comment]
In that case, I suggest removing the optimization for the local version.

**Comment by vangent** [comment]
> In that case, I suggest removing the optimization for the local version.
Why? It seems silly to create an fsnotifier that we know we're not going to use, and I think it actually makes the code more understandable since it points out that we're only going to use the notifier if we've already returned a value, and it hasn't changed since then.

**Pull request merged**